coderd/workspaceapps/proxy.go

```diff
@@ -0,0 +1,594 @@
+		// See https://github.com/coder/code-server/issues/241 for examples.
+		http.Redirect(rw, r, r.URL.Path+"/", http.StatusTemporaryRedirect)
+		return
+	}
+	if path == "/" && r.URL.RawQuery == "" && appURL.RawQuery != "" {
+		// If the application defines a default set of query parameters,
+		// we should always respect them. The reverse proxy will merge
+		// query parameters for server-side requests, but sometimes
+		// client-side applications require the query parameters to render
+		// properly. With code-server, this is the "folder" param.
+		r.URL.RawQuery = appURL.RawQuery
+		http.Redirect(rw, r, r.URL.String(), http.StatusTemporaryRedirect)
+		return
+	}
+
+	r.URL.Path = path
+	appURL.RawQuery = ""
+
+	proxy := httputil.NewSingleHostReverseProxy(appURL)
+	proxy.ErrorHandler = func(w http.ResponseWriter, r *http.Request, err error) {
+		site.RenderStaticErrorPage(rw, r, site.ErrorPageData{
+			Status:       http.StatusBadGateway,
+			Title:        "Bad Gateway",
+			Description:  "Failed to proxy request to application: " + err.Error(),
+			RetryEnabled: true,
+			DashboardURL: s.PrimaryAccessURL.String(),
+		})
+	}
+
+	conn, release, err := s.WorkspaceConnCache.Acquire(appToken.AgentID)
+	if err != nil {
+		site.RenderStaticErrorPage(rw, r, site.ErrorPageData{
+			Status:       http.StatusBadGateway,
+			Title:        "Bad Gateway",
+			Description:  "Could not connect to workspace agent: " + err.Error(),
+			RetryEnabled: true,
+			DashboardURL: s.PrimaryAccessURL.String(),
+		})
+		return
+	}
+	defer release()
+	proxy.Transport = conn.HTTPTransport()
+
+	// This strips the session token from a workspace app request.
+	cookieHeaders := r.Header.Values("Cookie")[:]
+	r.Header.Del("Cookie")
+	for _, cookieHeader := range cookieHeaders {
+		r.Header.Add("Cookie", httpapi.StripCoderCookies(cookieHeader))
+	}
+
+	// Convert canonicalized headers to their non-canonicalized counterparts.
+	// See the comment on `nonCanonicalHeaders` for more information on why this
+	// is necessary.
+	for k, v := range r.Header {
+		if n, ok := nonCanonicalHeaders[k]; ok {
+			r.Header.Del(k)
+			r.Header[n] = v
+		}
+	}
+
+	// end span so we don't get long lived trace data
+	tracing.EndHTTPSpan(r, http.StatusOK, trace.SpanFromContext(ctx))
+
+	proxy.ServeHTTP(rw, r)
+}
+
+// workspaceAgentPTY spawns a PTY and pipes it over a WebSocket.
+// This is used for the web terminal.
+//
+// @Summary Open PTY to workspace agent
+// @ID open-pty-to-workspace-agent
+// @Security CoderSessionToken
+// @Tags Agents
+// @Param workspaceagent path string true "Workspace agent ID" format(uuid)
+// @Success 101
+// @Router /workspaceagents/{workspaceagent}/pty [get]
+func (s *WorkspaceAppServer) workspaceAgentPTY(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	// TODO: Fix this later
+	// s.WebsocketWaitMutex.Lock()
+	// s.WebsocketWaitGroup.Add(1)
+	// s.WebsocketWaitMutex.Unlock()
+	// defer s.WebsocketWaitGroup.Done()
```

Comment:
This will cause leaks if merged before fixing this, right? We could just add a `Close` func to this server, which has it's own mutex and wait group for this.

Reply:
I can do that. 

Question, why do we need a mutex and a wait group? Aren't wait groups already thread safe?

This code in the close locks the mutex then waits.
https://github.com/coder/coder/blob/7e0ea1006f8c9707cf6619ae3509b0119ac57599/coderd/coderd.go#L809-L811

So if a new websocket comes in while we are closing, the new websocket conn gets block (trys to lock). Then the wait group finises, then the new websocket conn continues as the lock is freed.

I think this logic has some flaws, and just a waitgroup wouldn't fix all race conditions, but it keeps the same behavior and reduces some complexity.

Reply:
I've implemented it the exact same way as coderd does. I do agree that the mutex seems pointless and we can fix that if you want

Reply:
Is this crazy?
https://github.com/coder/coder/pull/7008